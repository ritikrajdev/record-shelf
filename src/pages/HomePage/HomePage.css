.sync-screen {
  flex: 1;
  flex-direction: column;
  display: flex;
  justify-content: center;
  align-items: center;
  row-gap: 1rem;
}

.sync-screen > p {
  font-size: 2rem;
  font-weight: 200;
}

.container {
  flex: 1;
  padding: 2rem 3rem;
  margin: 0 max(10vw, 2rem);
  word-wrap: normal;
  background-color: var(--light-gray-bg);
}

.container-bar {
  margin-bottom: 2rem;
  display: flex;
  justify-content: space-between;
}

.container-heading {
  font-size: 3.5rem;
}

.records {
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
  gap: 1rem;
}

@media screen and (max-width: 460px) {
  .sync-screen > p {
    font-size: 1.5rem;
  }

  .container-heading {
    font-size: 3rem;
  }

  .container {
    margin: 0;
    padding: 1rem;
  }
}
