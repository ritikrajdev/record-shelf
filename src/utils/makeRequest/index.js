import axios from 'axios';

export async function makeRequest(
  apiEndpoint = {},
  dyamicConfig = {},
  navigate
) {
  try {
    const response = await axios({
      ...dyamicConfig,
      ...apiEndpoint,
      headers: {
        Authorization: 'Bearer QWlzaHdhcnlhIE4=',
      },
    });

    return response.data;
  } catch (err) {
    navigate('/error');
  }
}
