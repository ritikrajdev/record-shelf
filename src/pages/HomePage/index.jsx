import React, { useState } from 'react';
import IconButton from '../../components/elements/IconButton';
import RoundedButton from '../../components/elements/RoundedButton';

import iconGenre from '../../assets/icons/icon-genre.svg';
import iconGrid from '../../assets/icons/icon-grid.svg';

import { useNavigate, useSearchParams } from 'react-router-dom';
import GenreIcon from '../../components/elements/GenreIcon';
import Record from '../../components/elements/Record';

import {
  GET_ALL_RECORDS,
  GET_LIKES,
  UPDATE_LIKES,
} from '../../constants/apiEndPoints';
import { makeRequest } from '../../utils/makeRequest';
import './HomePage.css';

export default function HomePage() {
  const [searchParams, setSearchParams] = useSearchParams();

  const [groupByGenre, setGroupByGenre] = useState(
    searchParams.get('groupBy') === 'genre'
  );
  const [records, setRecords] = useState(null);

  const navigate = useNavigate();

  const fetchAndUpdateRecords = async () => {
    const recordsDataWithoutLikes = (
      await makeRequest(GET_ALL_RECORDS, {}, navigate)
    ).data;
    const likesResponseData = await Promise.all(
      recordsDataWithoutLikes.map((record) =>
        makeRequest(GET_LIKES(record.id), {}, navigate)
      )
    );
    const likesData = likesResponseData.map((resData) => resData.data);
    const recordsData = recordsDataWithoutLikes.map((record, idx) => ({
      ...record,
      genre: record.genre.name,
      artist: record.artist.name,
      isLiked: likesData[idx].like,
      numLikes: likesData[idx].count,
    }));
    console.log(recordsData);
    setRecords(recordsData);
  };

  const toggleGroupBy = () => {
    if (groupByGenre) {
      setSearchParams({});
    } else {
      setSearchParams({ groupBy: 'genre' });
    }

    setGroupByGenre(!groupByGenre);
  };

  if (records === null) {
    return (
      <div className='sync-screen'>
        <p>{':(('}</p>
        <p>seems a bit empty here ...</p>
        <RoundedButton onClick={fetchAndUpdateRecords}>Sync</RoundedButton>
      </div>
    );
  }

  const groupedRecords = {};
  if (groupByGenre) {
    records.forEach((record) => {
      groupedRecords[record.genre] = groupedRecords[record.genre] ?? [];
      groupedRecords[record.genre].push(record);
    });
  }

  const toggleLike = async (recordId) => {
    const recordIndex = records.findIndex((record) => record.id === recordId);

    const isLiked = records[recordIndex].isLiked;
    try {
      const likedData = (
        await makeRequest(
          UPDATE_LIKES(recordId, {
            like: !isLiked,
          })
        )
      ).data;

      records[recordIndex].isLiked = likedData.like;
      records[recordIndex].numLikes = likedData.count;
      setRecords([...records]);
    } catch (e) {
      console.error(e);
    }
  };

  return (
    <div className='container' data-testid='container'>
      <div className='container-bar'>
        <h2 className='container-heading'>all songs</h2>
        <IconButton
          onClick={toggleGroupBy}
          iconSrc={groupByGenre ? iconGrid : iconGenre}
          width='2.5rem'
          height='2.5rem'
        />
      </div>

      {groupByGenre ? (
        Object.entries(groupedRecords).map(([genre, genreRecords], index) => {
          return (
            <div key={index}>
              <GenreIcon genre={genre.toLowerCase()} />
              <br />
              <div className='records'>
                {genreRecords.map((record) => (
                  <Record
                    key={record.id}
                    recordData={record}
                    onLikeClick={() => toggleLike(record.id)}
                  />
                ))}
              </div>
              <br />
              <br />
            </div>
          );
        })
      ) : (
        <div className='records'>
          {records.map((record) => (
            <Record
              key={record.id}
              recordData={record}
              onLikeClick={() => toggleLike(record.id)}
            />
          ))}
        </div>
      )}
    </div>
  );
}
