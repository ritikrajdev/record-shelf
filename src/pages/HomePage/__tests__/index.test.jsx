import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import HomePage from '..';
import {
  mockLikeGetResponse,
  mockRecordsGetResponse,
} from '../../../mocks/recordsResponse';
import { makeRequest } from '../../../utils/makeRequest';

jest.mock('../../../utils/makeRequest');

describe('HomePage', () => {
  it('renders', async () => {
    const screen = render(
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>
    );
    expect(screen.asFragment()).toMatchSnapshot();
  });

  it('should call makeRequest on sync button click and update data in container', async () => {
    makeRequest
      .mockResolvedValueOnce(mockRecordsGetResponse)
      .mockResolvedValue(mockLikeGetResponse);
    const screen = render(
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>
    );
    const syncButton = screen.getByText('Sync');
    fireEvent.click(syncButton);
    await waitFor(() => {
      expect(screen.getByTestId('container')).toBeTruthy();
    });
  });
});
