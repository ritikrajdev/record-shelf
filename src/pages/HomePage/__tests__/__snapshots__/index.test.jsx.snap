// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders 1`] = `
<DocumentFragment>
  <div
    class="sync-screen"
  >
    <p>
      :((
    </p>
    <p>
      seems a bit empty here ...
    </p>
    <button
      class="rounded-button"
    >
      Sync
    </button>
  </div>
</DocumentFragment>
`;
